diff --git a/src/kernel/thread.c b/src/kernel/thread.c
--- a/src/kernel/thread.c
+++ b/src/kernel/thread.c
@@ -1,7 +1,9 @@
   *--kernel_stack_ptr = 0; // r13
   *--kernel_stack_ptr = 0; // r14
   *--kernel_stack_ptr = 0; // r15
 
+  thread->rsp = (uint64_t)kernel_stack_ptr;
+
   klog(LOG_INFO, "thread_create_userspace: thread ID %d created successfully",
        thread->id);
   scheduler_add_thread(thread);
